tests/lib/prepare-utils.sh

```diff
@@ -0,0 +1,314 @@
+    # from when we build the debian package
+    add-apt-repository ppa:snappy-dev/image -y
+    apt update -qq
+    apt upgrade -yqq
+
+    # these are already installed in the lxd image which speeds things up, but they
+    # are missing in qemu and google images.
+    apt install initramfs-tools-core psmisc fdisk snapd mtools ovmf qemu-system-x86 sshpass whois openssh-server -yqq
+
+    # use the snapd snap explicitly
+    # TODO: since ubuntu-image ships it's own version of `snap prepare-image`, 
+    # should we instead install beta/edge snapd here and point ubuntu-image to this
+    # version of snapd?
+    # TODO: https://bugs.launchpad.net/snapd/+bug/1712808
+    # There is a bug in snapd that prevents udev rules from reloading in privileged containers
+    # with the following error message: 'cannot reload udev rules: exit status 1' when installing
+    # snaps. However it seems that retrying the installation fixes it
+    if ! snap install snapd; then
+        echo "FIXME: snapd install failed, retrying"
+        snap install snapd
+    fi
+    snap install ubuntu-image --classic
+
+    # install build-deps for ubuntu-core-initramfs
+    (
+        cd "$project_dir"
+        apt -y build-dep ./
+    )
+}
+
+download_core_initrd_snaps() {
+    local snap_branch="$1"
+
+    # get the model
+    curl -o ubuntu-core-amd64-dangerous.model https://raw.githubusercontent.com/snapcore/models/master/ubuntu-core-22-amd64-dangerous.model
+
+    # download neccessary images
+    snap download pc-kernel --channel=22/${snap_branch} --basename=upstream-pc-kernel
+    snap download pc --channel=22/${snap_branch} --basename=upstream-pc-gadget
+    snap download snapd --channel=${snap_branch} --basename=upstream-snapd
+    snap download core22 --channel=${snap_branch} --basename=upstream-core22
+}
+
+build_core_initrd() {
+    local project_dir="$1"
+    local current_dir="$(pwd)"
+    
+    # build the debian package of ubuntu-core-initramfs
+    (
+        cd "$project_dir"
+        DEB_BUILD_OPTIONS='nocheck testkeys' dpkg-buildpackage -tc -b -Zgzip
+
+        # save our debs somewhere safe
+        cp ../*.deb "$current_dir"
+    )
+}
+
+add_core_initrd_ssh_users() {
+    snapddir=/tmp/snapd-workdir
+    unsquashfs -d $snapddir upstream-snapd.snap
+
+    # inject systemd service to setup users and other tweaks for us
+    # these are copied from upstream snapd prepare.sh, slightly modified to not 
+    # extract snapd spread data from ubuntu-seed as we don't need all that here
+    cat > "$snapddir/lib/systemd/system/snapd.spread-tests-run-mode-tweaks.service" <<'EOF'
+[Unit]
+Description=Tweaks to run mode for spread tests
+Before=snapd.service
+Documentation=man:snap(1)
+
+[Service]
+Type=oneshot
+ExecStart=/usr/lib/snapd/snapd.spread-tests-run-mode-tweaks.sh
+RemainAfterExit=true
+
+[Install]
+WantedBy=multi-user.target
+EOF
+
+cat > "$snapddir/usr/lib/snapd/snapd.spread-tests-run-mode-tweaks.sh" <<'EOF'
```

Comment:
Should not this be indented?

Reply:
Really good catch, it should be indented!